In [1]:
import pandas as pd

In [2]:
# Load the full dataset
bank_dataset = pd.read_csv("A2-bank/bank-additional-full.csv", sep=";")

# Print the first few rows of the dataset
bank_dataset.head()


age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

The Bank Marketing dataset contains 41,188 instances and 21 features. Most of the features are categorical (object data type), with a few numerical ones.

There are 'unknown' values in six features: 'job', 'marital', 'education', 'default', 'housing', and 'loan'. The number of 'unknown' values in these features ranges from 80 to 8,597.

To handle the 'unknown' values and represent the categorical features as numerical data, we can perform the following steps:

    Imputation: Replace the 'unknown' values with the most frequent value in each feature. This is a simple imputation method that can work well when the number of 'unknown' values is relatively small compared to the size of the dataset.

    Encoding: Convert the categorical features into numerical form. One common way to do this is by using one-hot encoding, which creates binary (0 or 1) features for each category of each original feature.

Let's perform these preprocessing steps. After that, we'll split the dataset into a training set (first 80% patterns) and a test set (last 20% patterns), as per your instructions.

In [3]:
# Print the size of the dataset
print("Size of the dataset: ", bank_dataset.shape)

# Print the data types of the features
print("\nData types of the features:")
print(bank_dataset.dtypes)

# Count the number of "unknown" values in each column
unknown_counts = bank_dataset.apply(lambda x: x[x == 'unknown'].count(), axis=0)
print("\nNumber of 'unknown' values in each column:")
print(unknown_counts[unknown_counts > 0])


Size of the dataset:  (41188, 21)

Data types of the features:
age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

Number of 'unknown' values in each column:
job           330
marital        80
education    1731
default      8597
housing       990
loan          990
dtype: int64


The 'unknown' values have been replaced with the most frequent value in each feature, and the categorical features have been converted into numerical form using label encoding. Each unique category in each feature has been assigned a unique integer.

In [4]:
from sklearn.preprocessing import LabelEncoder

# Replace 'unknown' with the most frequent value in each column
for column in bank_dataset.columns:
    most_frequent = bank_dataset[column].mode()[0]
    bank_dataset[column] = bank_dataset[column].replace('unknown', most_frequent)

# Convert categorical features into numerical form using label encoding
for column in bank_dataset.columns:
    if bank_dataset[column].dtype == 'object':
        le = LabelEncoder()
        bank_dataset[column] = le.fit_transform(bank_dataset[column])

# Print the first few rows of the preprocessed dataset
bank_dataset.head()


age  job  marital  education  default  housing  loan  contact  month  \
0   56    3        1          0        0        0     0        1      6   
1   57    7        1          3        0        0     0        1      6   
2   37    7        1          3        0        1     0        1      6   
3   40    0        1          1        0        0     0        1      6   
4   56    7        1          3        0        0     1        1      6   

   day_of_week  ...  campaign  pdays  previous  poutcome  emp.var.rate  \
0            1  ...         1    999         0         1           1.1   
1            1  ...         1    999         0         1           1.1   
2            1  ...         1    999         0         1           1.1   
3            1  ...         1    999         0         1           1.1   
4            1  ...         1    999         0         1           1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0          93.994          -36.4      4.857       5191.0  0  
1          93.994          -36.4      4.857       5191.0  0  
2          93.994          -36.4      4.857       5191.0  0  
3          93.994          -36.4      4.857       5191.0  0  
4          93.994          -36.4      4.857       5191.0  0  

[5 rows x 21 columns]

Now, let's split the dataset into a training set (first 80% of patterns) and a test set (last 20% of patterns). We'll use these sets to train and test the SVM, BP, and MLR models, as we did with the Ring datasets. 

In [5]:
# Split the dataset into features (X) and target variable (y)
X_bank = bank_dataset.drop('y', axis=1)
y_bank = bank_dataset['y']

# Calculate the split index
split_idx = int(0.8 * len(X_bank))

# Split the dataset into a training set and a test set
X_train_bank = X_bank[:split_idx]
y_train_bank = y_bank[:split_idx]
X_test_bank = X_bank[split_idx:]
y_test_bank = y_bank[split_idx:]

# Print the sizes of the training set and test set
print("Size of the training set: ", X_train_bank.shape[0])
print("Size of the test set: ", X_test_bank.shape[0])


Size of the training set:  32950
Size of the test set:  8238


The Bank Marketing dataset has been successfully split into a training set of 32,950 instances and a test set of 8,238 instances, according to the 80-20 split specified in your instructions.

Let's proceed with applying the Support Vector Machines (SVM) and Multiple Linear Regression (MLR) methods to the dataset. 

For SVM, we'll start with the 'rbf' kernel and 'C' and 'gamma' parameters set to 1 and 'scale', respectively. 
For MLR, we'll use the default parameters. 
We'll evaluate each model on the test set, reporting the classification error, confusion matrix, and ROC AUC score.

In [6]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score

# Function to calculate classification error
def classification_error(y_true, y_pred):
    return 100 * (sum(y_true != y_pred) / len(y_true))

# Function to apply SVM to the given dataset and print the classification error, confusion matrix, and ROC AUC score
def apply_svm(X_train, y_train, X_test, y_test, cv=True):
    """
    Apply SVM to the given dataset and print the classification error, confusion matrix, and ROC AUC score.
    The function supports skipping cross-validation by setting cv=3.
    """
    # Create an SVM classifier with 'rbf' kernel
    svm = SVC(kernel='rbf', C=1, gamma='scale', probability=True, random_state=0)
    
    if cv:
        # Set a grid for the 'C' and 'gamma' parameters
        param_grid = {'C': [0.1, 1, 10], 'gamma': [1, 0.1, 0.01, 0.001]}

        # Perform cross-validation to find best parameters
        grid = GridSearchCV(svm, param_grid, cv=3)
        grid.fit(X_train, y_train)

        # Print best parameters
        print("Best parameters: ", grid.best_params_)

        # Use the SVM classifier with best parameters
        svm = SVC(kernel='rbf', C=grid.best_params_['C'], gamma=grid.best_params_['gamma'], probability=True, random_state=0)
    
    # Train the SVM classifier
    svm.fit(X_train, y_train)

    # Predict the targets for the test set
    y_pred = svm.predict(X_test)
    y_prob = svm.predict_proba(X_test)[:,1]

    # Print the classification error
    print("\nClassification error: {:.2f}%".format(classification_error(y_test, y_pred)))

    # Print the confusion matrix
    print("\nConfusion matrix:\n", confusion_matrix(y_test, y_pred))

    # Calculate the ROC AUC score
    roc_auc = roc_auc_score(y_test, y_prob)
    print("\nROC AUC score: {:.2f}".format(roc_auc))

    return svm
 
# Apply SVM to the Bank Marketing dataset with default parameters (without cross-validation)
print("Bank Marketing Dataset")
svm_bank = apply_svm(X_train_bank, y_train_bank, X_test_bank, y_test_bank, cv=False)



Bank Marketing Dataset

Classification error: 30.83%

Confusion matrix:
 [[5694    4]
 [2536    4]]

ROC AUC score: 0.76


In [7]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score
from sklearn.linear_model import LinearRegression

In [8]:
# Function to apply MLR and convert predictions into class labels
def apply_mlr(X_train, y_train, X_test, y_test):
    # Create a linear regression model
    mlr = LinearRegression()

    # Train the model
    mlr.fit(X_train, y_train)

    # Make predictions
    y_pred_cont = mlr.predict(X_test)

    # Convert continuous predictions into class labels using a threshold
    threshold = 0.5
    y_pred = (y_pred_cont > threshold).astype("int32")

    # Calculate and print accuracy
    accuracy = np.mean(y_pred == y_test)
    print("Accuracy: ", accuracy)

    print("Classification report:")
    print(classification_report(y_test, y_pred))
    print("Confusion matrix:")
    print(confusion_matrix(y_test, y_pred))
    print("ROC AUC Score: ", roc_auc_score(y_test, y_pred))

    return mlr


In [9]:
# Apply Multiple Linear Regression (MLR) to the Bank Marketing dataset
print("\nMultiple Linear Regression (MLR) - Bank Marketing Dataset")
mlr_bank = apply_mlr(X_train_bank, y_train_bank, X_test_bank, y_test_bank)


Multiple Linear Regression (MLR) - Bank Marketing Dataset
Accuracy:  0.6996843894149065
Classification report:
              precision    recall  f1-score   support

           0       0.70      0.99      0.82      5698
           1       0.70      0.05      0.09      2540

    accuracy                           0.70      8238
   macro avg       0.70      0.52      0.45      8238
weighted avg       0.70      0.70      0.59      8238

Confusion matrix:
[[5648   50]
 [2424  116]]
ROC AUC Score:  0.518447141281787
